src/internal_modules/roc_core/heap_arena.cpp

```diff
@@ -42,13 +44,24 @@ size_t HeapArena::num_allocations() const {
 void* HeapArena::allocate(size_t size) {
     num_allocations_++;
 
-    Chunk* chunk = (Chunk*)malloc(sizeof(Chunk) + size);
+    size_t total_size = (sizeof(Chunk) + sizeof(ChunkCanary) + size + sizeof(ChunkCanary)
+                         + sizeof(AlignMax) - 1)
+        / sizeof(AlignMax) * sizeof(AlignMax);
+    size_t size_padding = AlignOps::align_max(size) - size;
```

Comment:
There is no need to apply alignment or padding here.

In pool, single slab was containing multiple slots placed continuously. If the size of a slot is not max-aligned, next slot will start from non-aligned address, and in result its data will be non-aligned.

In heap arena, we don't have this problem. Each chunk does its own call to malloc(). malloc() always returns max-aligned memory, and Chunk::data is of type AlignMax; because of this, data is guaranteed to be aligned.

Reply:
So the new code is:

```cpp
    size_t size_aligned = AlignOps::align_max(size);
    size_t total_size =
        sizeof(ChunkHeader) + sizeof(ChunkCanary) + size_aligned + sizeof(ChunkCanary);
    size_t size_padding = size_aligned - size;
```

But why do you think we need align_max() and size_padding?

My understanding is that we need just:

```cpp
size_t total_size =
        sizeof(ChunkHeader) + sizeof(ChunkCanary) + size + sizeof(ChunkCanary);
```

Reply:
My thinking...
Since we need size_aligned, in order to calculate the size_padding, why not use that in the total_size.

Also, I'm not sure, but some malloc libraries might actually track this to see if we are accessing outside the size requested?

Reply:
Why do we need size_aligned and size_padding?

My point is that padding is not needed in HeapArena at all.

Instead of:

```cpp
MemoryOps::prepare_canary(canary_after, size_padding + sizeof(ChunkCanary));
```

we can use:

```
MemoryOps::prepare_canary(canary_after, size + sizeof(ChunkCanary));
```

Reply:
> Also, I'm not sure, but some malloc libraries might actually track this to see if we are accessing outside the size requested?

That's true. There are a lot of different tools. E.g. on MSVC you can enable guard pages around each allocations. On glibc there is MALLOC_CHECK_, and in clang and gcc there are sanitizers (which we support via --sanitizers scons option).

These checks are toolchain-dependent and enabled conditionally. Often they should be enabled at build time and can either affect performance and/or require dependencies to be built with same flags.

In this issue we implement minimalistic portable version of such check that will be always enabled, regardless of platform, build flags, etc, and hopefully will help to detect certain class of bugs.

And we will also continue using more serious checks as well. For example, on CI we have jobs that run tests under sanitizers and valgrind.

Reply:
> Instead of: [...]

Sorry, the code I provided is wrong. Here is what I mean:

```cpp
    const size_t chunk_size =
        sizeof(ChunkHeader) + sizeof(ChunkCanary) + size + sizeof(ChunkCanary);

    ChunkHeader* chunk = (ChunkHeader*)malloc(chunk_size);

    char* canary_before = (char*)chunk->data;
    char* memory = (char*)chunk->data + sizeof(ChunkCanary);
    char* canary_after = (char*)chunk->data + sizeof(ChunkCanary) + size;

    MemoryOps::prepare_canary(canary_before, sizeof(ChunkCanary));
    MemoryOps::poison_before_use(memory, size);
    MemoryOps::prepare_canary(canary_after, sizeof(ChunkCanary));
```

i.e., there is no need to pad `size` to make it max-aligned. Because of reasons I mentioned here: https://github.com/roc-streaming/roc-toolkit/pull/597#discussion_r1348795259

Reply:
In PoolImpl, after the data, we use prepare_canary() and check_canary() with padding bytes + canary-size. So, for non-max-aligned data, we'll always check a few more byets. I was going for something similar here.

But, let me know. I'll do what you suggested.

Reply:
Yeah,

In pool, we need to add padding to object to ensure that **slot size** of multiple of max-align bytes.

If beginning of first slot is max-aligned, and size of each slot is multiple of max-align bytes, it is then guaranteed that beginning of **each** slot is max-aligned.

In other words, we need to pad objects because we split continuous chunk into multiple slots and need to align each of them.

However, in arena, there are no multiple slots. We need to ensure that beginning of user data is max-aligned. But we don't care if size of user data is multiple of max-align bytes, because we don't place anything after user data.

Well, we place canary guard there, but we don't care about its alignment.

We need max-aligning for user data because we don't know its structure. Maybe user data is begins with uint64_t, which requires to be 8-byte aligned on some CPUs, and otherwise attempt to access it will cause SIGBUS.

But our canary guard is accessed via MemoryOps; it interprets canary as bytes and does not place any requirements on alignment of canary.

Hence, in arena we can place trailing canary right after user data and don't care about padding. This will simplify implementation a bit and we'll avoid unnecessary calculations.